bindings/rust/bench/src/s2n_tls.rs

```diff
@@ -0,0 +1,184 @@
+            errno::set_errno(errno::Errno(libc::EWOULDBLOCK));
+            -1
+        } else {
+            info!("\t- received {len}");
+            len as _
+        }
+    }
+
+    fn create_config(mode: Mode) -> Config {
+        let mut builder = Builder::new();
+        builder.set_security_policy(&DEFAULT_TLS13).unwrap();
+
+        match mode {
+            Mode::Server => builder
+                .load_pem(
+                    read_to_bytes(SERVER_CERT_CHAIN_PATH).as_slice(),
+                    read_to_bytes(SERVER_KEY_PATH).as_slice(),
+                )
+                .unwrap(),
+            Mode::Client => builder
+                .trust_pem(read_to_bytes(CA_CERT_PATH).as_slice())
+                .unwrap()
+                .set_verify_host_callback(HostNameHandler {
+                    hostname: "localhost",
+                })
+                .unwrap(),
+        };
+
+        builder.build().unwrap()
+    }
+
+    /// Set up connections with config and custom IO
+    fn init_conn(&mut self, mode: Mode) {
+        let c_to_s_ptr = &mut self.c_to_s_buf as &mut Buffer as *mut Buffer as *mut c_void;
+        let s_to_c_ptr = &mut self.s_to_c_buf as &mut Buffer as *mut Buffer as *mut c_void;
+        let (read_ptr, write_ptr, config, conn);
+
+        match mode {
+            Mode::Client => {
+                read_ptr = s_to_c_ptr;
+                write_ptr = c_to_s_ptr;
+                config = &self.c_config;
+                conn = &mut self.c_conn;
+            }
+            Mode::Server => {
+                read_ptr = c_to_s_ptr;
+                write_ptr = s_to_c_ptr;
+                config = &self.s_config;
+                conn = &mut self.s_conn;
+            }
+        }
+
+        conn.set_blinding(Blinding::SelfService) // no blinding so benchmark time is accurate
+            .unwrap()
+            .set_config(config.clone())
+            .unwrap()
+            .set_send_callback(Some(Self::send_cb))
+            .unwrap()
+            .set_receive_callback(Some(Self::recv_cb))
+            .unwrap();
+        unsafe {
+            conn.set_send_context(write_ptr)
+                .unwrap()
+                .set_receive_context(read_ptr)
+                .unwrap();
+        }
+    }
+
+    /// Handshake step for one connection
+    fn handshake_conn(&mut self, mode: Mode) {
+        let (conn, handshaked);
+        match mode {
+            Mode::Client => {
+                info!("Client: ");
+                conn = &mut self.c_conn;
+                handshaked = &mut self.c_handshaked;
+            }
+            Mode::Server => {
+                info!("Server: ");
+                conn = &mut self.s_conn;
+                handshaked = &mut self.s_handshaked;
+            }
+        }
+        if let Ready(res) = conn.poll_negotiate() {
+            res.unwrap();
+            *handshaked = true;
+        } else {
+            *handshaked = false;
+        }
+    }
+}
+
+impl TlsImpl for S2nTls {
```

Comment:
nit `TlsImpl` feels like a kind of odd name? Perhaps `BenchHarness` or `TlsHarness`, or something like that.